### DISCUSSION TITLE
Gen Vx Mask Boundaries

### DISCUSSION THREAD

**Discussion by scole-mo** [dtcenter/METplus#1533] [Configuration] [answered]
Hi all,

I've been working with gen vx mask to replicate the current area masking function of the operational verification system at the Met Office (VER). I've been finding when using the same coordinates used in VER I get fewer grid points included in the gen vx mask generated areas. To account for this I applied an extension of 0.1 degrees to coordinates used - this made the grid point total in the gen vx mask area equal to that in VER.

I originally posed this question at a Met Office/ MET developers catch up towards the end of last year and this was suspected to be a precision issue. I've since moved on to work on a different resolution grid (1.5 degree instead of 2.5) and found the 0.1 degree extension doesn't make the number of grid points equal for VER and MET. Instead, a 0.6 degree extension has been needed. I was wondering if anyone had any thoughts on what could be causing this? Is it still likely to be a precision issue, or more of a logic issue?

I've attached some slides that show how I extended an area by 0.1 and the extensions I've tried.

Just to mention as well, I've found the same results when making masks using type ‘poly’ and a combination of type ‘lat’  and type ‘lon’

Many thanks in advance, 

Seb

[MET Developers Slide.pptx](https://github.com/dtcenter/METplus/files/8334720/MET.Developers.Slide.pptx)

**Comment by j-opatz** [comment, marked as answer]
Hi Seb,

And thank you for your question. I'm going to tag @JohnHalleyGotway as he will be able to answer this question more definitively. He's currently out of the office, but thank you for your patience in the meantime.

I'm not familiar with the previous discussions on this, but I also ran into something similar with some masking work I previously did. Like you, I found that some of the grid points would be excluded even though I had defined the poly file to be on the lat/lon lines that I wanted. Using the example of area 142 you provided in your slides, I thought that using 70 lat N would mean that grid points along the 70 degree latitude northern border would be included in my verification area. Likewise for the other borders. But I was finding that those grid points were falling outside my verification mask, leading to a lower count of included points.

When I discussed the issue with John HG, your suspicion of rounding was what John HG confirmed - limiting the poly definitions to ones accuracy (in your case) and trying to compare that to a float (or however the grid point values are stored) can cause comparison issues that the code sees the two values as different and discards those grid points as outside your masking region. 

As you found out, the way around that was to "nudge" the boundaries a bit, so that the grid points you want included in your area are "beyond rounding doubt" within the boundaries. That means that areas that are defined in a .poly file should be .2 or .3 in whatever direction away from the masking region.

The part that puzzles me is why you had to increase the masking limits to 0.6 rather than 0.1 on a finer resolution grid. This may still be an issue of rounding, though.

**Reply by scole-mo** [reply]
Thanks for this @j-opatz. Interesting to hear you've found the same. Another thing I tried to get around this was generating a mask outside of MET but I've found this not to work either. I've attached a mask I made of the same area, 142 on 2.5 degree grid.
[142custom_25_deg.zip](https://github.com/dtcenter/METplus/files/8362242/142custom_25_deg.zip)

**Reply by JohnHalleyGotway** [reply]
@scole-mo and @j-opatz, I read through this discussion to get up to speed.

Let me clarify a couple of details. Here's the list of Point-Stat masking options:
```
mask = {
   grid  = [];
   poly  = [];
   sid   = [];
   llpnt = [];
}
```

The regions specified in the **grid** and **poly** options are applied to the gridded forecast data. All forecast values for grid points included in those regions are retained, while values for grid points outside those regions are set to bad data. Along the boundaries, we have valid data values on the inside and bad data values on the outside. Then the user-specified interpolation is performed. For nearest neighbor, if the "closest" grid point contains valid data, it's used. If it contains bad data, it's not. For bilinear interpolation, all 4 nearby grid points would need to contain valid data. For other interpolations methods and widths, the vld_thresh setting would indicate whether or not enough valid data is present for an interpolated value to be computed. So if @j-opatz was using nearest neighbor and @scole-mo was using bilinear, that would explain the need for a "larger buffer".

The **sid** and **llpnt** options are applied to the observation points themselves. The station ID list (**sid**) is pretty self-explanatory. The observation is used only if its station ID appears in the user-specified list. The **llpnt** option applies thresholds to the Lat/Lon values of observations themselves. The observation is used only if its Lat/Lon values meet the specified thresholds.

It sounds to me like you'd want to use the **llpnt** option to have finer control over which points are included. The downside is that this will only work well for rectangular regions.

Please search for "llpnt" in [this section](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html?highlight=llpnt#mask) of the User's Guide to see some examples.

Hopefully this helps clarify the behavior you two were seeing.

**Reply by JohnHalleyGotway** [reply]
Just wanted to follow up on this discussion. @scole-mo have your questions been addressed? If so, please select one of the responses as being a good "Answer" and click the "Mark as answer" button. Thanks!

**Reply by scole-mo** [reply]
Thanks for this and apologies for the slow reply @JohnHalleyGotway. I haven't actually got this to work as yet, I've been trying llpnt but with gridstat, should this still work? Also was llpnt introduced in a specific version of MET?

**Reply by JohnHalleyGotway** [reply]
The mask.llpnt option applies only when processing point observations. The Point-Stat and Ensemble-Stat tools process point observations. That's why llpnt appears as options in the [PointStatConfig_default](https://github.com/dtcenter/MET/blob/205fc9476aebac5bbfc18dfdb2fea38bd1959531/met/data/config/PointStatConfig_default#L210) and [EnsembleStatConfig_default](https://github.com/dtcenter/MET/blob/205fc9476aebac5bbfc18dfdb2fea38bd1959531/met/data/config/EnsembleStatConfig_default#L233) files.

The Grid-Stat tool does not process point observations. It does a grid-to-grid comparison and that's why the llpnt option does NOT appear in its [GridStatConf_default](https://github.com/dtcenter/MET/blob/205fc9476aebac5bbfc18dfdb2fea38bd1959531/met/data/config/GridStatConfig_default#L119) file.

Below, I've copied and pasted an explanation from [this section](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html?highlight=llpnt) of the MET User's Guide. Hope that helps clarify...

The masking logic for processing point observations in Point-Stat and Ensemble-Stat fall into two categories. The “sid” and “llpnt” options apply directly to the point observations. Only those observations for the specified station id’s are included in the “sid” masks. Only those observations meeting the latitude and longitude threshold criteria are included in the “llpnt” masks.

The “grid” and “poly” mask options are applied to the grid points of the verification domain. Each grid point is determined to be inside or outside the masking region. When processing point observations, their latitude and longitude values are rounded to the nearest grid point of the verification domain. If the nearest grid point is inside the mask, that point observation is included in the mask.

**Reply by JohnHalleyGotway** [reply]
Just wanted to followup on this discussion. @scole-mo do you have any more questions about the masking options?

**Reply by scole-mo** [reply]
Thanks @JohnHalleyGotway this has been really helpful!
